File: src/prototype/web/src/input/layerB_defense.ts
// PURE — Layer B (defender) per docs/design/input_system_defense_v1.md §B.
//
// Transforms an InputFrame into a DefenseIntent. Structurally parallel to
// input/layerB.ts but with different semantics for each control:
//   LS → weight_forward / weight_lateral
//   RS → base support zone (trigger held) or cut direction (bumper edge)
//   L/R TRIGGER → base pressure (continuous)
//   L/R BUMPER → cut attempt (edge)
//   BTN_BASE → recovery hold
//   BTN_RELEASE → base release
//   BTN_BREATH → breath recovery
//   BTN_RESERVED → pass commit

import {
  BASE_ZONE_DIRECTIONS,
  ZERO_TOP_BASE,
  type BaseZone,
  type DefenseDiscreteIntent,
  type DefenseIntent,
  type TopBaseIntent,
  type TopHipIntent,
} from "./intent_defense.js";
import { ButtonBit, type InputFrame } from "./types.js";

export type LayerBDefenseState = Readonly<{
  lastBaseZone: BaseZone | null;
}>;

export const INITIAL_LAYER_B_DEFENSE_STATE: LayerBDefenseState = Object.freeze({
  lastBaseZone: null,
});

// Same hysteresis logic as attacker Layer B, just operating on BaseZone.
const ZONE_SELECT_COS_THRESHOLD = Math.cos((30 * Math.PI) / 180);
const RS_MAGNITUDE_THRESHOLD = 0.2;

export function transformLayerBDefense(
  frame: InputFrame,
  prev: LayerBDefenseState,
): { intent: DefenseIntent; nextState: LayerBDefenseState } {
  const hip = computeTopHipIntent(frame);
  const { base, nextZone } = computeTopBaseIntent(frame, prev.lastBaseZone);
  const discrete = computeDefenseDiscreteIntents(frame);

  return {
    intent: Object.freeze({ hip, base, discrete }),
    nextState: Object.freeze({ lastBaseZone: nextZone }),
  };
}

export function computeTopHipIntent(frame: InputFrame): TopHipIntent {
  return Object.freeze({
    weight_forward: frame.ls.y,
    weight_lateral: frame.ls.x,
  });
}

// Base zones are used for trigger-held "support hand" placement. During
// a bumper edge, RS is repurposed for cut direction — base zone selection
// is suppressed that frame to avoid ambiguity.
export function computeTopBaseIntent(
  frame: InputFrame,
  lastZone: BaseZone | null,
): { base: TopBaseIntent; nextZone: BaseZone | null } {
  const bumperEdge =
    (frame.button_edges & (ButtonBit.L_BUMPER | ButtonBit.R_BUMPER)) !== 0;
  const anyTrigger = frame.l_trigger > 0 || frame.r_trigger > 0;
  const rsMag = Math.hypot(frame.rs.x, frame.rs.y);

  let nextZone: BaseZone | null = lastZone;

  // RS only feeds base-zone selection while a trigger is held AND no
  // bumper is edging this frame.
  if (!bumperEdge && anyTrigger && rsMag >= RS_MAGNITUDE_THRESHOLD) {
    const nx = frame.rs.x / rsMag;
    const ny = frame.rs.y / rsMag;
    const best = pickBestBaseZone(nx, ny);
    if (lastZone === null || best.zone === lastZone) {
      nextZone = best.zone;
    } else if (best.dot >= ZONE_SELECT_COS_THRESHOLD) {
      nextZone = best.zone;
    }
  } else if (!anyTrigger) {
    // No trigger held → nothing to place. Drop the zone so the next press
    // starts fresh.
    nextZone = null;
  }

  const l_hand_target = frame.l_trigger > 0 ? nextZone : null;
  const r_hand_target = frame.r_trigger > 0 ? nextZone : null;

  const base: TopBaseIntent =
    l_hand_target === null && r_hand_target === null
      ? ZERO_TOP_BASE
      : Object.freeze({
          l_hand_target,
          l_base_pressure: frame.l_trigger,
          r_hand_target,
          r_base_pressure: frame.r_trigger,
        });

  return { base, nextZone };
}

function pickBestBaseZone(nx: number, ny: number): { zone: BaseZone; dot: number } {
  let bestDot = -Infinity;
  let bestZone: BaseZone = BASE_ZONE_DIRECTIONS[0]!.zone;
  for (const { zone, dir } of BASE_ZONE_DIRECTIONS) {
    const d = nx * dir.x + ny * dir.y;
    if (d > bestDot) {
      bestDot = d;
      bestZone = zone;
    }
  }
  return { zone: bestZone, dot: bestDot };
}

export function computeDefenseDiscreteIntents(
  frame: InputFrame,
): ReadonlyArray<DefenseDiscreteIntent> {
  const out: DefenseDiscreteIntent[] = [];
  if ((frame.button_edges & ButtonBit.L_BUMPER) !== 0) {
    out.push({ kind: "CUT_ATTEMPT", side: "L", rs: frame.rs });
  }
  if ((frame.button_edges & ButtonBit.R_BUMPER) !== 0) {
    out.push({ kind: "CUT_ATTEMPT", side: "R", rs: frame.rs });
  }
  if ((frame.buttons & ButtonBit.BTN_BASE) !== 0) {
    out.push({ kind: "RECOVERY_HOLD" });
  }
  if ((frame.button_edges & ButtonBit.BTN_RELEASE) !== 0) {
    out.push({ kind: "BASE_RELEASE_ALL" });
  }
  if ((frame.button_edges & ButtonBit.BTN_BREATH) !== 0) {
    out.push({ kind: "BREATH_START" });
  }
  if ((frame.button_edges & ButtonBit.BTN_RESERVED) !== 0) {
    out.push({ kind: "PASS_COMMIT", rs: frame.rs });
  }
  if ((frame.button_edges & ButtonBit.BTN_PAUSE) !== 0) {
    out.push({ kind: "PAUSE" });
  }
  return Object.freeze(out);
}

File: src/prototype/unity/Assets/BJJSimulator/Runtime/Input/LayerBDefense.cs
// Ported 1:1 from src/prototype/web/src/input/layerB_defense.ts.
// PURE — Layer B (defender): InputFrame → DefenseIntent.
// Reference: docs/design/input_system_defense_v1.md §B.
//
// Structurally parallel to LayerB.cs but with defender-specific semantics.

namespace BJJSimulator
{
    // -------------------------------------------------------------------------
    // Base-zone direction table (§B.2 top-side controls)
    // -------------------------------------------------------------------------

    public struct BaseZoneDir
    {
        public BaseZone Zone;
        public Vec2     Dir;
    }

    // -------------------------------------------------------------------------
    // Per-player state
    // -------------------------------------------------------------------------

    public struct LayerBDefenseState
    {
        public BaseZone LastBaseZone;

        public static readonly LayerBDefenseState Initial =
            new LayerBDefenseState { LastBaseZone = BaseZone.None };
    }

    // -------------------------------------------------------------------------
    // Pure transform
    // -------------------------------------------------------------------------

    public static class LayerBDefenseOps
    {
        private const float S = 0.7071068f; // 1/√2
        public static readonly BaseZoneDir[] BaseZoneDirections = new BaseZoneDir[]
        {
            new BaseZoneDir { Zone = BaseZone.Chest,  Dir = new Vec2( 0f,  1f) },
            new BaseZoneDir { Zone = BaseZone.Hip,    Dir = new Vec2( 0f, -1f) },
            new BaseZoneDir { Zone = BaseZone.KneeL,  Dir = new Vec2(-S,  -S) },
            new BaseZoneDir { Zone = BaseZone.KneeR,  Dir = new Vec2( S,  -S) },
            new BaseZoneDir { Zone = BaseZone.BicepL, Dir = new Vec2(-S,   S) },
            new BaseZoneDir { Zone = BaseZone.BicepR, Dir = new Vec2( S,   S) },
        };

        private const float ZoneSelectCosThreshold = 0.866f; // cos(30°)
        private const float RsMagnitudeThreshold   = 0.2f;

        // -----------------------------------------------------------------------

        public static (DefenseIntent Intent, LayerBDefenseState NextState) Transform(
            InputFrame frame,
            LayerBDefenseState prev)
        {
            var hip               = ComputeTopHipIntent(frame);
            var (base_, nextZone) = ComputeTopBaseIntent(frame, prev.LastBaseZone);
            var discrete          = ComputeDefenseDiscreteIntents(frame);

            return (
                new DefenseIntent { Hip = hip, Base = base_, Discrete = discrete },
                new LayerBDefenseState { LastBaseZone = nextZone }
            );
        }

        public static TopHipIntent ComputeTopHipIntent(InputFrame frame) =>
            new TopHipIntent { WeightForward = frame.Ls.Y, WeightLateral = frame.Ls.X };

        public static (TopBaseIntent Base, BaseZone NextZone) ComputeTopBaseIntent(
            InputFrame frame, BaseZone lastZone)
        {
            bool bumperEdge = (frame.ButtonEdges & (ButtonBit.LBumper | ButtonBit.RBumper)) != 0;
            bool anyTrigger = frame.LTrigger > 0f || frame.RTrigger > 0f;
            float rsMag     = frame.Rs.Magnitude;
            BaseZone next   = lastZone;

            if (!bumperEdge && anyTrigger && rsMag >= RsMagnitudeThreshold)
            {
                float nx = frame.Rs.X / rsMag;
                float ny = frame.Rs.Y / rsMag;
                var best = PickBestBaseZone(nx, ny);
                if (lastZone == BaseZone.None || best.Zone == lastZone)
                    next = best.Zone;
                else if (best.Dot >= ZoneSelectCosThreshold)
                    next = best.Zone;
            }
            else if (!anyTrigger)
            {
                next = BaseZone.None;
            }

            BaseZone lTarget = frame.LTrigger > 0f ? next : BaseZone.None;
            BaseZone rTarget = frame.RTrigger > 0f ? next : BaseZone.None;

            TopBaseIntent base_;
            if (lTarget == BaseZone.None && rTarget == BaseZone.None)
            {
                base_ = TopBaseIntent.Zero;
            }
            else
            {
                base_ = new TopBaseIntent
                {
                    LHandTarget   = lTarget,
                    LBasePressure = frame.LTrigger,
                    RHandTarget   = rTarget,
                    RBasePressure = frame.RTrigger,
                };
            }

            return (base_, next);
        }

        public static DefenseDiscreteIntent[] ComputeDefenseDiscreteIntents(InputFrame frame)
        {
            var buf = new System.Collections.Generic.List<DefenseDiscreteIntent>(4);
            if ((frame.ButtonEdges & ButtonBit.LBumper) != 0)
                buf.Add(new DefenseDiscreteIntent
                {
                    Kind    = DefenseDiscreteIntentKind.CutAttempt,
                    CutSide = HandSide.L,
                    Rs      = frame.Rs,
                });
            if ((frame.ButtonEdges & ButtonBit.RBumper) != 0)
                buf.Add(new DefenseDiscreteIntent
                {
                    Kind    = DefenseDiscreteIntentKind.CutAttempt,
                    CutSide = HandSide.R,
                    Rs      = frame.Rs,
                });
            if ((frame.Buttons & ButtonBit.BtnBase) != 0)
                buf.Add(new DefenseDiscreteIntent { Kind = DefenseDiscreteIntentKind.RecoveryHold });
            if ((frame.ButtonEdges & ButtonBit.BtnRelease) != 0)
                buf.Add(new DefenseDiscreteIntent { Kind = DefenseDiscreteIntentKind.BaseReleaseAll });
            if ((frame.ButtonEdges & ButtonBit.BtnBreath) != 0)
                buf.Add(new DefenseDiscreteIntent { Kind = DefenseDiscreteIntentKind.BreathStart });
            if ((frame.ButtonEdges & ButtonBit.BtnReserved) != 0)
                buf.Add(new DefenseDiscreteIntent
                {
                    Kind = DefenseDiscreteIntentKind.PassCommit,
                    Rs   = frame.Rs,
                });
            if ((frame.ButtonEdges & ButtonBit.BtnPause) != 0)
                buf.Add(new DefenseDiscreteIntent { Kind = DefenseDiscreteIntentKind.Pause });
            return buf.ToArray();
        }

        // -----------------------------------------------------------------------

        private struct BestBase { public BaseZone Zone; public float Dot; }

        private static BestBase PickBestBaseZone(float nx, float ny)
        {
            float    bestDot = float.NegativeInfinity;
            BaseZone best    = BaseZone.None;
            foreach (var e in BaseZoneDirections)
            {
                float d = nx * e.Dir.X + ny * e.Dir.Y;
                if (d > bestDot) { bestDot = d; best = e.Zone; }
            }
            return new BestBase { Zone = best, Dot = bestDot };
        }
    }
}
